dbms/tests/integration/test_storage_s3/test.py

```diff
@@ -1,159 +1,168 @@
+
+    assert csv_data == get_s3_file_content(cluster, filename)
+
+
+# Test put and get with S3 server redirect.
+def test_put_get_with_redirect(cluster):
+    # type: (ClickHouseCluster) -> None
+
+    instance = cluster.instances["dummy"]  # type: ClickHouseInstance
+    table_format = "column1 UInt32, column2 UInt32, column3 UInt32"
+    values = "(1, 1, 1), (1, 1, 1), (11, 11, 11)"
+    values_csv = "1,1,1\n1,1,1\n11,11,11\n"
+    filename = "test.csv"
+    query = "insert into table function s3('http://{}:{}/{}/{}', 'CSV', '{}') values {}".format(
+        cluster.minio_redirect_host, cluster.minio_redirect_port, cluster.minio_bucket, filename, table_format, values)
     run_query(instance, query)
 
-    query = "select *, column1*column2*column3 from s3('http://{}:{}/{}/test.csv', 'CSV', '{}')".format(started_cluster.mock_host, started_cluster.preserving_data_port, started_cluster.bucket, format)
+    assert values_csv == get_s3_file_content(cluster, filename)
+
+    query = "select *, column1*column2*column3 from s3('http://{}:{}/{}/{}', 'CSV', '{}')".format(
+        cluster.minio_redirect_host, cluster.minio_redirect_port, cluster.minio_bucket, filename, table_format)
     stdout = run_query(instance, query)
+
     assert list(map(str.split, stdout.splitlines())) == [
         ["1", "1", "1", "1"],
         ["1", "1", "1", "1"],
         ["11", "11", "11", "1331"],
     ]
-    data = get_communication_data(started_cluster)
-    received_data = data["received_data"]
-    assert received_data[-1].decode() == "1,1,1\n1,1,1\n11,11,11\n"
-
-
-def test_multipart_put(started_cluster):
-    instance = started_cluster.instances["dummy"]
-    format = "column1 UInt32, column2 UInt32, column3 UInt32"
-
-    put_communication_data(started_cluster, "=== Multipart test ===")
-    long_data = [[i, i+1, i+2] for i in range(100000)]
-    long_values = "".join([ "{},{},{}\n".format(x,y,z) for x, y, z in long_data ])
-    put_query = "insert into table function s3('http://{}:{}/{}/test.csv', 'CSV', '{}') format CSV".format(started_cluster.mock_host, started_cluster.multipart_preserving_data_port, started_cluster.bucket, format)
-    run_query(instance, put_query, stdin=long_values, settings={'s3_min_upload_part_size': 1000000})
-    data = get_communication_data(started_cluster)
-    assert "multipart_received_data" in data
-    received_data = data["multipart_received_data"]
-    assert received_data[-1].decode() == "".join([ "{},{},{}\n".format(x, y, z) for x, y, z in long_data ])
-    assert 1 < data["multipart_parts"] < 10000
+
+
+# Test multipart put.
+def test_multipart_put(cluster):
+    # type: (ClickHouseCluster) -> None
+
+    instance = cluster.instances["dummy"]  # type: ClickHouseInstance
+    table_format = "column1 UInt32, column2 UInt32, column3 UInt32"
+    long_data = [[i, i + 1, i + 2] for i in range(100000)]
```

Comment:
We need to generate more than 5MB of data here and it should be done in more predictable way than just `100000`, see `get_random_string()`.

Reply:
Fixed.